google-cloud-bigquery/lib/google/cloud/bigquery/convert.rb

```diff
@@ -378,9 +378,50 @@ def self.derive_source_format path
         #   is nil.
         def self.millis_to_time time_millis
           return nil unless time_millis
-          time_millis = Integer time_millis
-          time_secs = time_millis / 1000.0
-          ::Time.at time_secs
+          ::Time.at Rational(time_millis, 1000)
+        end
+
+        ##
+        # @private
+        #
+        # Converts a Ruby Time object to a primitive time value in milliseconds.
+        #
+        # @return [Integer, nil] The primitive time value in milliseconds, or
+        #   nil if the given argument is nil.
+        def self.time_to_millis time_obj
+          return nil unless time_obj
+          (time_obj.to_i * 1000) + (time_obj.nsec / 1000000)
+        end
+
+        ##
+        # @private
+        #
+        # Converts an array of Google::Apis::BigqueryV2::StandardSqlField to
```

Comment:
I'm not confident this simplification of the type representation makes sense, but BQML models are fairly restrictive in what types are currently allowed as feature and label columns.  With routines, we'll need to expose the StandardSqlStructType/StandardSQLField/StandardSQLDataType as proper resources as it's the types we'll use for exposing function arguments/return values.

Reply:
Spanner has a very similar structure to describe fields, including nested arrays and structs. In Spanner users sometimes need to specify this type information when it can't be inferred from the data correctly, like when specifying `NULL` while using query parameters. We use this particular structure to so users can specify the values using Ruby primitives, which are easier to create than the verbose objects as defined by protobuf/discovery. We thought that since this structure was useful in Spanner it would also be useful in BigQuery. Is this explanation persuasive at all?

Reply:
Ahh, I missed the nuance about swapping representations.  In the BQML models usage, we're basically just informing users which columns are supplied to do a prediction, and which columns contain the prediction results.  For this, having the hash with the simplified name/type should be sufficient.

Future work is coming where we'll want users to be able to express things using the backend type representation, so I just wanted to ensure we're able to express the same level of complexity.

Reply:
We could use the same approach for this data as we use for the training_runs data. Meaning, instead of returning this particular primitive representation we could return the unserialized values or the apiary structured objects.

Reply:
I've updated the PR and removed the coercion to a Ruby Hash. These methods now return the deserialized JSON from the API, same as `#training_runs`.